lib/mlir/Dialect/Clift/IR/CliftAttributes.cpp

```diff
@@ -102,11 +141,10 @@ mlir::LogicalResult FieldAttr::verify(EmitErrorType EmitError,
                                       uint64_t Offset,
                                       clift::ValueType ElementType,
                                       llvm::StringRef Name) {
-  if (auto Definition = mlir::dyn_cast<DefinedType>(ElementType)) {
-    if (mlir::isa<FunctionTypeAttr>(Definition.getElementType()))
-      return EmitError() << "Underlying type of field attr cannot be a "
-                            "function type";
-  }
+  if (not isObjectType(ElementType))
+    return EmitError() << "Struct and union field types must be object types.";
```

Comment:
:thinking: this makes it impossible to have struct/union fields that are primitives, enums, pointers, which is wrong.

The only forbidden types should be incomplete types (like you properly check below), and zero-sized types, that you also check below.

Reply:
And function types. Also scalar tuple types actually, which I'm not checking for. I'm borrowing C++ terminology here. Note that *object type* is not *class type* (which includes structs and unions). *Object type* is what you might think of as value type. It's anything that you can have an object of and it includes primitive types, enums, pointers.